oeps/architectural-decisions/oep-0065-frontend-composability.rst

```diff
@@ -0,0 +1,492 @@
+     - 3.7k
+     - Ultimately should be replaced with @reduxjs/toolkit. This document isn't making a judgment on the utility of redux, just noting that it is currently in use in the vast majority of MFEs.
+   * - regenerator-runtime
+     - 6.6k
+     -
+
+Total size: ~2,087.2k unzipped
+
+Shared dependency advantages
+----------------------------
+
+- Encourages runtime consistency across all types of dependencies. If our MFEs use the same version, we significantly reduce the cognitive load and context switching involved in working with multiple MFEs, cut down on visual inconsistencies at runtime, etc.
+- Reduces runtime bundle size. We'll ship far less code to the client across a user's browsing session.
+
+Caveat: "build time" and "dependency maintenance"
+-------------------------------------------------
+
+Note that "build time" and "dependency maintenance" are not mentioned in the advantages above.  The reasons for this have to do with how shared dependencies are resolved for modules at runtime, and are described in more detail in the `Module federation`_ section below.
+
+At a high level, even if MFEs share dependencies, we want to preserve the ability for them to "fall back" to their own version of a shared dependency if a version already loaded on the page is incompatible with their own.  To do this, each MFE builds and deploys its own version of all its dependencies in case dependency resolution determines they're needed.
+
+This means their build has to take time to bundle them, and the repository still needs the dependencies present in the package.json file.  Sharing dependencies doesn't help us much w/r/t build time and dependency maintenance.
+
+Approach
+========
+
+We intend to enable runtime module loading and shared dependencies via `Webpack module federation <https://webpack.js.org/concepts/module-federation>`_. More information on module federation beyond its webpack implementation can be found on its `dedicated site at module-federation.io <https://module-federation.io/>`_.
+
+Further, we believe that we also need an approach to maintaining and enforcing dependency consistency across MFEs to realize the benefits of shared dependencies. MFEs need a compatible version of the shared dependency to be available, otherwise they need to fall back to their own version. The process, tooling, and/or code organization necessary to provide that consistency is not something module federation can help with and needs to be addressed separately.
+
+Module Federation
+-----------------
+
+Module federation is implemented as a `plugin for Webpack <https://webpack.js.org/plugins/module-federation-plugin/>`_ that enables micro-frontends to be composed into the same page at runtime even if they're built separately and independently deployed. The pieces being composed are "modules". It lets us configure which dependencies should be shared between modules on a page and what modules a particular frontend exposes to be loaded by other frontends.
```

Comment:
[frontend-baby question] is `module` here just a webpack-y term for what we call MFEs? Or put another way, would there be `modules` in the Open edX ecosystem that _weren't_ MFEs?

Reply:
I think that would be a good addition, as a glossary entry or footnote. I like what we did with [OEP-55](https://github.com/openedx/open-edx-proposals/blob/master/oeps/processes/oep-0055-proc-project-maintainers.rst) with footnotes.

But maybe my question still stands - is an MFE itself a module? could it be multiple modules? Could we (and would we want to) compose non-MFE modules?